python/libsharp/libsharp.pyx

```diff
@@ -230,11 +232,54 @@ cdef class alm_info:
             raise NotInitializedError()
         return sharp_alm_count(self.ainfo)
 
+    def mval(self):
+        if self.ainfo == NULL:
+            raise NotInitializedError()
+        return np.asarray(<int[:self.ainfo.nm]> self.ainfo.mval)
+
+    def mvstart(self):
+        if self.ainfo == NULL:
+            raise NotInitializedError()
+        return np.asarray(<long[:self.ainfo.nm]> self.ainfo.mvstart)
+
     def __dealloc__(self):
         if self.ainfo != NULL:
             sharp_destroy_alm_info(self.ainfo)
         self.ainfo = NULL
 
+    def almxfl(self, np.ndarray[double, ndim=3, mode='c'] alm, np.ndarray[double, ndim=2, mode='c'] fl):
+        """Multiply Alm by a Ell based array
+
+
+        Parameters
+        ----------
+        alm : np.ndarray
+            input alm, 3 dimensions = (different signal x polarizations x lm-ordering)
+        fl : np.ndarray
+            either 1 dimension, e.g. gaussian beam, or 2 dimensions e.g. a polarized beam
+
+        Returns
+        -------
+        None, it modifies alms in-place
+
+        """
+        mvstart = 0
+        has_multiple_beams = alm.shape[2] > 1 and fl.shape[1] > 1
+        for m in self.mval():
```

Comment:
This is going to be inefficient as Cython can't optimize looping over this array. Accessing self.ainfo.mvstart directly would be faster by far.

Reply:
thanks @dagss why can't `cython` optimize it?

Reply:
Because it's a Python object. Cython doesn't do full program analysis; if something turns into a Python object (which it does when it is returned from mval()) then it will continue to be slow.

(Don't know if you knew but I was the one who added NumPy support to Cython :) but that *could* have changed lately, haven't been involved for some years, but it seems unlikely as it would be a quite sophisticated optimization.)

Reply:
@dagss mval() returns a numpy array, so that should not be a problem. Anyway, I'll inspect with the cython tools and report back.

I didn't know about you implementing that, that's awesome!

Reply:
It is not known compile time that mval() returns a NumPy error. YOU know it, but Cython does not.

To speed it up you would have to make a "cdef" NumPy array variable declaration inside the calling function and assign the result of calling mval() to it. (Also last time I checked nobody had written an optimization to loop over NumPy arrays even if the type is known compile time, you had to loop over and index i and index into the array on each iteration to make it fast -- but somebody may have implemented this since... )

Since it is not the inner loop I am not sure how serious it is though.

Reply:
I meant NumPy ARRAY not error of course